Add Unit Tests

Fl1s opened the pull request:
Add a bunch of tests for directories:
/config/
/core/http/
/core/https/ -> I had serious troubles with this one 
/exception/
/utils/

<img width="746" alt="result" src="https://github.com/user-attachments/assets/5cd3d39d-7543-4842-b8de-222a6e274f0a">

I'm having some difficulties with HttpServerListenerThreadTest and HttpsServerListenerThreadTest in particular. 
Namely:

HttpServerListenerThreadTest -
`java.net.BindException: Address already in use`
<img width="455" alt="test1" src="https://github.com/user-attachments/assets/68796f71-e40d-4dd3-a43b-c46777af7edb">
I forgot how to release the port after a test, and therefore need to fix the @AfterEach block to do this.

HttpsServerListenerThreadTest - I fundamentally did not make any changes to the core logic, writing only code for tests. Therefore, I did not create getters and setters methods. I also could not test the private method createSSLServerSocket() because of its access modifier. And lastly, I think we need method to close serverListenerThread.

Guhapriya01 commented:
@Fl1s, 

I will refactor the `createSSLServerSocket()` method and add getters to the server listener threads so that you can test them. Also, we can change the `tearDown` method to help fix the `BindException` you’re seeing.

Could you please try the updated `tearDown` method below and see if it works?

```java
@AfterEach
public void tearDown() throws IOException {
    if (serverListener.isAlive()) {
        serverListener.interrupt(); // Stop the thread if it's running
    }
    if (serverSocket != null && !serverSocket.isClosed()) {
        serverSocket.close(); 
    }
    if (clientSocket != null && !clientSocket.isClosed()) {
        clientSocket.close(); 
    }
}
```

Let me know how it goes!

Fl1s commented:
Unfortunately, the address is still in use:
<img width="452" alt="tests" src="https://github.com/user-attachments/assets/ef5e797e-91b5-445e-a4d3-c5c0149388a0">

Guhapriya01 commented:
> Gm @Guhapriya01! Unfortunately, the address is still in use: <img alt="tests" width="452" src="https://private-user-images.githubusercontent.com/73421069/376130209-ef5e797e-91b5-445e-a4d3-c5c0149388a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg4OTAxMDQsIm5iZiI6MTcyODg4OTgwNCwicGF0aCI6Ii83MzQyMTA2OS8zNzYxMzAyMDktZWY1ZTc5N2UtOTFiNS00NDVlLWE0ZDMtYzVjMDE0OTM4OGEwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDE0VDA3MTAwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjZjE2MzYzM2NhYThjODYyZDhkNTVmODQ0YjBmNTU3MWE5ZjRkZTU4MmY3MzY0MmNjNzgyZWM0YWZkNGZlYjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HmOdAJxu0xQuT1eiWNkExNBoZz4IhbLH0TMiWmR6Zco">


@Fl1s, I see that you’re still encountering the `BindException`. Sometimes, adding a short delay before the next test starts can help resolve this issue.

Could you try adding this line at the end of the `tearDown` method?

```java
Thread.sleep(100); 
```

Let me know if this helps!

Fl1s commented:
Nah, that didn't help, but I figured out where the bug was hiding:
`serverListener = new HttpServerListenerThread(0, "/webroot")`
It turned out that when I specified port 8080, it was full. As soon as I changed it to any free port(0), all tests passed!
<img width="463" alt="tests" src="https://github.com/user-attachments/assets/17a570bb-ac9e-47ab-9845-3f84d29858a7">

Guhapriya01 commented:
@Fl1s, I’ve changed the access modifier of the createSSLServerSocket method and added getters. Could you make the necessary changes to the HttpsServerListenerThreadTest accordingly?

Fl1s commented:
> @Fl1s, I’ve changed the access modifier of the createSSLServerSocket method and added getters. Could you make the necessary changes to the HttpsServerListenerThreadTest accordingly?

Done 😼

Guhapriya01 commented:
> Done 😼

Since there were some changes since the last code coverage report, could you please share the updated one?

Fl1s commented:
> > Done 😼
>  
> Since there were some changes since the last code coverage report, could you please share the updated one?

Check my last commit `Add more tests + optimized it`

Guhapriya01 commented:
> Check my last commit `Add more tests + optimized it`

@Fl1s, everything looks good! I would just like to see the updated code coverage. Could you please share it?

Fl1s commented:
Oh, right
<img width="739" alt="tests" src="https://github.com/user-attachments/assets/0506cf59-5cdd-4dcd-b64a-8df56bfe9746">
Just In HttpsServerListenerThreadTest, you will need to insert SSLConfiguration
<img width="492" alt="" src="https://github.com/user-attachments/assets/87fb4534-4d6e-4244-8b37-f047499ed9fc">

Guhapriya01 commented:
@Fl1s,

I noticed an issue while running the tests for the `HttpsServerListenerThread` class.

![error message](https://github.com/user-attachments/assets/7bb0ac91-30e6-45dc-a7dd-ef854c111c1d)

Are you aware of this issue? It seems like we might need to add `@AfterEach` to deregister the static mock. Am I missing anything in my setup, or could you address this error in your code?

Thanks!

Fl1s commented:
Try this one:
```java
@AfterEach
public void tearDown() {
    if (mockedSSLFactory != null) {
        mockedSSLFactory.close();
    }
}
```
If everything works fine, I'll upload it to a new commit

Guhapriya01 commented:
@Fl1s,

Yes! This worked:

```java
@AfterEach
public void tearDown() {
    mockedConfigManager.close();
    mockedSSLFactory.close();  
}
```

Also, please change the `IOException` to `RuntimeException` in the `testSSLServerSocketCreationFailure` method. Since `SSLServerSocketFactory.getDefault()` does not throw checked exceptions, we should mock it to throw an unchecked exception (like `RuntimeException`).

Here’s the error I encountered:

```
org.mockito.exceptions.base.MockitoException: 
Checked exception is invalid for this method!
Invalid: java.io.IOException: Failed to create SSL Server Socket
	at com.httpserver.core.https.HttpsServerListenerThreadTest.testSSLServerSocketCreationFailure(HttpsServerSocketCreationTest.java:90)
```

Fl1s commented:
That's kind of it now...Whew

Guhapriya01 commented on src/test/java/com/httpserver/core/https/HttpsServerListenerThreadTest.java:
```diff
@@ -24,111 +25,89 @@ public class HttpsServerListenerThreadTest {
-            );
-        }*/
+    public void testSSLServerSocketCreationFailure() throws RuntimeException {
+        mockedSSLFactory.when(SSLServerSocketFactory::getDefault).thenThrow(new IOException("Failed to create SSL Server Socket"));
```
Change IOException to RuntimeException

Guhapriya01 commented:
@Fl1s,

I really appreciate your contribution to this project. Great work! Your efforts in enhancing our test coverage are invaluable, and these updates will help ensure the reliability of our project.

Thanks again for your contribution! ❤️

Fl1s commented:
🥳
